src/plugins/vis_type_pie/public/pie_component.tsx

```diff
@@ -0,0 +1,338 @@
+  const handleSliceClick = useCallback(
+    (
+      clickedLayers: LayerValue[],
+      bucketColumns: Array<Partial<BucketColumns>>,
+      visData: Datatable,
+      splitChartDimension?: DatatableColumn,
+      splitChartFormatter?: FieldFormat
+    ): void => {
+      const data = getFilterClickData(
+        clickedLayers,
+        bucketColumns,
+        visData,
+        splitChartDimension,
+        splitChartFormatter
+      );
+      const event = {
+        name: 'filterBucket',
+        data: { data },
+      };
+      props.fireEvent(event);
+    },
+    [props]
+  );
+
+  // handles legend action event data
+  const getLegendActionEventData = useCallback(
+    (visData: Datatable) => (series: SeriesIdentifier): ClickTriggerEvent | null => {
+      const data = getFilterEventData(visData, series);
+
+      return {
+        name: 'filterBucket',
+        data: {
+          negate: false,
+          data,
+        },
+      };
+    },
+    []
+  );
+
+  const handleLegendAction = useCallback(
+    (event: ClickTriggerEvent, negate = false) => {
+      props.fireEvent({
+        ...event,
+        data: {
+          ...event.data,
+          negate,
+        },
+      });
+    },
+    [props]
+  );
+
+  const toggleLegend = useCallback(() => {
+    setShowLegend((value) => {
+      const newValue = !value;
+      props.uiState?.set('vis.legendOpen', newValue);
+      return newValue;
+    });
+  }, [props.uiState]);
+
+  useEffect(() => {
+    setShowLegend(props.visParams.addLegend);
+    props.uiState?.set('vis.legendOpen', props.visParams.addLegend);
+  }, [props.uiState, props.visParams.addLegend]);
+
+  const setColor = useCallback(
+    (newColor: string | null, seriesLabel: string | number) => {
+      const colors = props.uiState?.get('vis.colors') || {};
+      if (colors[seriesLabel] === newColor || !newColor) {
+        delete colors[seriesLabel];
+      } else {
+        colors[seriesLabel] = newColor;
+      }
+      props.uiState?.setSilent('vis.colors', null);
+      props.uiState?.set('vis.colors', colors);
+      props.uiState?.emit('reload');
+    },
+    [props.uiState]
+  );
+
+  const { visData, visParams, services, syncColors } = props;
+
+  function getSliceValue(d: Datum, metricColumn: DatatableColumn) {
+    if (typeof d[metricColumn.id] === 'number' && d[metricColumn.id] !== 0) {
+      return d[metricColumn.id];
+    }
+    return Number.EPSILON;
+  }
+
+  // formatters
+  const metricFieldFormatter = services.fieldFormats.deserialize(
+    visParams.dimensions.metric.format
+  );
+  const splitChartFormatter = visParams.dimensions.splitColumn
+    ? services.fieldFormats.deserialize(visParams.dimensions.splitColumn[0].format)
+    : visParams.dimensions.splitRow
+    ? services.fieldFormats.deserialize(visParams.dimensions.splitRow[0].format)
+    : undefined;
+  const percentFormatter = services.fieldFormats.deserialize({
+    id: 'percent',
+    params: {
+      pattern: '0.[00]%',
+    },
+  });
```

Comment:
I don't like that the formatter is hardcoded here, this is an area where you've chosen to give users less control than Lens. Would be nice to allow the decimal places to be controlled by the user.

Reply:
Correct! Done